File: compiler/forget/packages/babel-plugin-react-forget/src/HIR/FindContextIdentifiers.ts
import type { NodePath } from "@babel/traverse";
import type * as t from "@babel/types";
import { CompilerError } from "../CompilerError";
import { GeneratedSource } from "./HIR";

type FindContextIdentifierState = {
  inLambda: number;
  currentLambda: Array<
    | NodePath<t.FunctionDeclaration>
    | NodePath<t.FunctionExpression>
    | NodePath<t.ArrowFunctionExpression>
  >;
  contextIdentifiers: Set<t.Identifier>;
};

export function findContextIdentifiers(
  func: NodePath<t.Function>
): Set<t.Identifier> {
  const state: FindContextIdentifierState = {
    inLambda: 0,
    currentLambda: [],
    contextIdentifiers: new Set(),
  };

  func.traverse<FindContextIdentifierState>(
    {
      FunctionDeclaration: {
        enter(
          fn: NodePath<t.FunctionDeclaration>,
          state: FindContextIdentifierState
        ): void {
          state.currentLambda.push(fn);
        },
        exit(
          fn: NodePath<t.FunctionDeclaration>,
          state: FindContextIdentifierState
        ): void {
          state.currentLambda.pop();
        },
      },
      FunctionExpression: {
        enter(
          fn: NodePath<t.FunctionExpression>,
          state: FindContextIdentifierState
        ): void {
          state.currentLambda.push(fn);
        },
        exit(
          _fn: NodePath<t.FunctionExpression>,
          state: FindContextIdentifierState
        ): void {
          state.currentLambda.pop();
        },
      },

      ArrowFunctionExpression: {
        enter(
          fn: NodePath<t.ArrowFunctionExpression>,
          state: FindContextIdentifierState
        ): void {
          state.currentLambda.push(fn);
        },
        exit(
          _fn: NodePath<t.ArrowFunctionExpression>,
          state: FindContextIdentifierState
        ): void {
          state.currentLambda.pop();
        },
      },
      AssignmentExpression(
        path: NodePath<t.AssignmentExpression>,
        state: FindContextIdentifierState
      ): void {
        const currentLambda = state.currentLambda.at(-1);
        if (currentLambda) {
          const left = path.get("left");
          handleAssignment(currentLambda, state.contextIdentifiers, left);
        }
      },
    },
    state
  );
  return state.contextIdentifiers;
}

function handleAssignment(
  currentLambda:
    | NodePath<t.FunctionDeclaration>
    | NodePath<t.FunctionExpression>
    | NodePath<t.ArrowFunctionExpression>,
  contextIdentifiers: Set<t.Identifier>,
  lvalPath: NodePath<t.LVal>
): void {
  // Find all reassignments to identifiers declared outside of currentLambda
  // This closely follows destructuring assignment assumptions and logic in BuildHIR
  const lvalNode = lvalPath.node;
  switch (lvalNode.type) {
    case "Identifier": {
      const path = lvalPath as NodePath<t.Identifier>;
      const name = path.node.name;
      const ownBinding = path.scope.getBinding(name);
      const bindingAboveLambdaScope =
        currentLambda.scope.parent.getBinding(name);

      if (ownBinding != null && ownBinding === bindingAboveLambdaScope) {
        contextIdentifiers.add(ownBinding.identifier);
      }
      break;
    }
    case "ArrayPattern": {
      const path = lvalPath as NodePath<t.ArrayPattern>;
      for (const element of path.get("elements")) {
        if (nonNull(element)) {
          handleAssignment(currentLambda, contextIdentifiers, element);
        }
      }
      break;
    }
    case "ObjectPattern": {
      const path = lvalPath as NodePath<t.ObjectPattern>;
      for (const property of path.get("properties")) {
        if (property.isObjectProperty()) {
          const valuePath = property.get("value");
          if (!valuePath.isLVal()) {
            CompilerError.invariant(
              `[FindContextIdentifiers] Expected object property value to be an LVal, got: ${valuePath.type}`,
              valuePath.node.loc ?? GeneratedSource
            );
          }
          handleAssignment(currentLambda, contextIdentifiers, valuePath);
        } else {
          if (!property.isRestElement()) {
            CompilerError.invariant(
              `[FindContextIdentifiers] Invalid assumptions for babel types.`,
              property.node.loc ?? GeneratedSource
            );
          }
          handleAssignment(currentLambda, contextIdentifiers, property);
        }
      }
      break;
    }
    case "AssignmentPattern": {
      const path = lvalPath as NodePath<t.AssignmentPattern>;
      const left = path.get("left");
      handleAssignment(currentLambda, contextIdentifiers, left);
      break;
    }
    case "RestElement": {
      const path = lvalPath as NodePath<t.RestElement>;
      handleAssignment(currentLambda, contextIdentifiers, path.get("argument"));
      break;
    }
    case "MemberExpression": {
      // Interior mutability (not a reassign)
      break;
    }
    default: {
      CompilerError.todo(
        `[FindContextIdentifiers] Cannot handle Object destructuring assignment target ${lvalNode.type}`,
        lvalNode.loc ?? GeneratedSource
      );
    }
  }
}

function nonNull<T extends NonNullable<t.Node>>(
  t: NodePath<T | null>
): t is NodePath<T> {
  return t.node != null;
}

File: compiler/crates/react_compiler_lowering/src/find_context_identifiers.rs
//! Rust equivalent of the TypeScript `FindContextIdentifiers` pass.
//!
//! Determines which bindings need StoreContext/LoadContext semantics by
//! walking the AST with scope tracking to find variables that cross
//! function boundaries.

use std::collections::HashMap;
use std::collections::HashSet;

use react_compiler_ast::expressions::*;
use react_compiler_ast::patterns::*;
use react_compiler_ast::scope::*;
use react_compiler_ast::statements::FunctionDeclaration;
use react_compiler_ast::visitor::AstWalker;
use react_compiler_ast::visitor::Visitor;

use crate::FunctionNode;

#[derive(Default)]
struct BindingInfo {
    reassigned: bool,
    reassigned_by_inner_fn: bool,
    referenced_by_inner_fn: bool,
}

struct ContextIdentifierVisitor<'a> {
    scope_info: &'a ScopeInfo,
    /// Stack of inner function scopes encountered during traversal.
    /// Empty when at the top level of the function being compiled.
    function_stack: Vec<ScopeId>,
    binding_info: HashMap<BindingId, BindingInfo>,
}

impl<'a> ContextIdentifierVisitor<'a> {
    fn push_function_scope(&mut self, start: Option<u32>, node_id: Option<u32>) {
        let scope = node_id
            .and_then(|nid| self.scope_info.resolve_scope_by_node_id(nid))
            .or_else(|| start.and_then(|s| self.scope_info.node_to_scope.get(&s).copied()));
        if let Some(scope) = scope {
            self.function_stack.push(scope);
        }
    }

    fn pop_function_scope(&mut self, start: Option<u32>, node_id: Option<u32>) {
        let has_scope = node_id
            .and_then(|nid| self.scope_info.resolve_scope_by_node_id(nid))
            .or_else(|| start.and_then(|s| self.scope_info.node_to_scope.get(&s).copied()));
        if has_scope.is_some() {
            self.function_stack.pop();
        }
    }

    fn check_captured_reference(&mut self, start: Option<u32>, node_id: Option<u32>) {
        let binding_id = if let Some(nid) = node_id {
            match self.scope_info.resolve_reference_by_node_id(nid) {
                Some(id) => id,
                None => return,
            }
        } else {
            let start = match start {
                Some(s) => s,
                None => return,
            };
            match self.scope_info.reference_to_binding.get(&start) {
                Some(&id) => id,
                None => return,
            }
        };
        let &fn_scope = match self.function_stack.last() {
            Some(s) => s,
            None => return,
        };
        let binding = &self.scope_info.bindings[binding_id.0 as usize];
        if is_captured_by_function(self.scope_info, binding.scope, fn_scope) {
            let info = self.binding_info.entry(binding_id).or_default();
            info.referenced_by_inner_fn = true;
        }
    }

    fn handle_reassignment_identifier(&mut self, name: &str, current_scope: ScopeId) {
        if let Some(binding_id) = self.scope_info.get_binding(current_scope, name) {
            let info = self.binding_info.entry(binding_id).or_default();
            info.reassigned = true;
            if let Some(&fn_scope) = self.function_stack.last() {
                let binding = &self.scope_info.bindings[binding_id.0 as usize];
                if is_captured_by_function(self.scope_info, binding.scope, fn_scope) {
                    info.reassigned_by_inner_fn = true;
                }
            }
        }
    }
}

impl<'ast> Visitor<'ast> for ContextIdentifierVisitor<'_> {
    fn enter_function_declaration(&mut self, node: &'ast FunctionDeclaration, _: &[ScopeId]) {
        self.push_function_scope(node.base.start, node.base.node_id);
    }
    fn leave_function_declaration(&mut self, node: &'ast FunctionDeclaration, _: &[ScopeId]) {
        self.pop_function_scope(node.base.start, node.base.node_id);
    }
    fn enter_function_expression(&mut self, node: &'ast FunctionExpression, _: &[ScopeId]) {
        self.push_function_scope(node.base.start, node.base.node_id);
    }
    fn leave_function_expression(&mut self, node: &'ast FunctionExpression, _: &[ScopeId]) {
        self.pop_function_scope(node.base.start, node.base.node_id);
    }
    fn enter_arrow_function_expression(
        &mut self,
        node: &'ast ArrowFunctionExpression,
        _: &[ScopeId],
    ) {
        self.push_function_scope(node.base.start, node.base.node_id);
    }
    fn leave_arrow_function_expression(
        &mut self,
        node: &'ast ArrowFunctionExpression,
        _: &[ScopeId],
    ) {
        self.pop_function_scope(node.base.start, node.base.node_id);
    }
    fn enter_object_method(&mut self, node: &'ast ObjectMethod, _: &[ScopeId]) {
        self.push_function_scope(node.base.start, node.base.node_id);
    }
    fn leave_object_method(&mut self, node: &'ast ObjectMethod, _: &[ScopeId]) {
        self.pop_function_scope(node.base.start, node.base.node_id);
    }

    fn enter_identifier(&mut self, node: &'ast Identifier, _scope_stack: &[ScopeId]) {
        self.check_captured_reference(node.base.start, node.base.node_id);
    }

    fn enter_jsx_identifier(
        &mut self,
        node: &'ast react_compiler_ast::jsx::JSXIdentifier,
        _scope_stack: &[ScopeId],
    ) {
        self.check_captured_reference(node.base.start, node.base.node_id);
    }

    fn enter_assignment_expression(
        &mut self,
        node: &'ast AssignmentExpression,
        scope_stack: &[ScopeId],
    ) {
        let current_scope = scope_stack
            .last()
            .copied()
            .unwrap_or(self.scope_info.program_scope);
        walk_lval_for_reassignment(self, &node.left, current_scope);
    }

    fn enter_update_expression(&mut self, node: &'ast UpdateExpression, scope_stack: &[ScopeId]) {
        if let Expression::Identifier(ident) = node.argument.as_ref() {
            let current_scope = scope_stack
                .last()
                .copied()
                .unwrap_or(self.scope_info.program_scope);
            self.handle_reassignment_identifier(&ident.name, current_scope);
        }
    }
}

/// Recursively walk an LVal pattern to find all reassignment target identifiers.
fn walk_lval_for_reassignment(
    visitor: &mut ContextIdentifierVisitor<'_>,
    pattern: &PatternLike,
    current_scope: ScopeId,
) {
    match pattern {
        PatternLike::Identifier(ident) => {
            visitor.handle_reassignment_identifier(&ident.name, current_scope);
        }
        PatternLike::ArrayPattern(pat) => {
            for element in &pat.elements {
                if let Some(el) = element {
                    walk_lval_for_reassignment(visitor, el, current_scope);
                }
            }
        }
        PatternLike::ObjectPattern(pat) => {
            for prop in &pat.properties {
                match prop {
                    ObjectPatternProperty::ObjectProperty(p) => {
                        walk_lval_for_reassignment(visitor, &p.value, current_scope);
                    }
                    ObjectPatternProperty::RestElement(p) => {
                        walk_lval_for_reassignment(visitor, &p.argument, current_scope);
                    }
                }
            }
        }
        PatternLike::AssignmentPattern(pat) => {
            walk_lval_for_reassignment(visitor, &pat.left, current_scope);
        }
        PatternLike::RestElement(pat) => {
            walk_lval_for_reassignment(visitor, &pat.argument, current_scope);
        }
        PatternLike::MemberExpression(_) => {
            // Interior mutability - not a variable reassignment
        }
    }
}

/// Check if a binding declared at `binding_scope` is captured by a function at `function_scope`.
/// Returns true if the binding is declared above the function (in the parent scope or higher).
fn is_captured_by_function(
    scope_info: &ScopeInfo,
    binding_scope: ScopeId,
    function_scope: ScopeId,
) -> bool {
    let fn_parent = match scope_info.scopes[function_scope.0 as usize].parent {
        Some(p) => p,
        None => return false,
    };
    if binding_scope == fn_parent {
        return true;
    }
    // Walk up from fn_parent to see if binding_scope is an ancestor
    let mut current = scope_info.scopes[fn_parent.0 as usize].parent;
    while let Some(scope_id) = current {
        if scope_id == binding_scope {
            return true;
        }
        current = scope_info.scopes[scope_id.0 as usize].parent;
    }
    false
}

/// Build a set of `(BindingId, position)` pairs that are declaration sites
/// in `reference_to_binding`, not true references. Uses node-ID comparison
/// when available (from `ref_node_id_to_binding` + `declaration_node_id`),
/// falling back to position comparison otherwise.
fn build_declaration_positions(scope_info: &ScopeInfo) -> HashSet<(BindingId, u32)> {
    let mut result = HashSet::new();

    if !scope_info.ref_node_id_to_binding.is_empty() {
        // Node-ID path: match entries in ref_node_id_to_binding against
        // each binding's declaration_node_id. Then record the corresponding
        // position from declaration_start for use in the position-keyed loop.
        for (&ref_node_id, &binding_id) in &scope_info.ref_node_id_to_binding {
            let binding = &scope_info.bindings[binding_id.0 as usize];
            if binding.declaration_node_id == Some(ref_node_id) {
                if let Some(decl_start) = binding.declaration_start {
                    result.insert((binding_id, decl_start));
                }
            }
        }
    } else {
        // Position fallback: no node-IDs available (OXC/SWC path)
        for binding in &scope_info.bindings {
            if let Some(decl_start) = binding.declaration_start {
                result.insert((binding.id, decl_start));
            }
        }
    }

    result
}

/// Find context identifiers for a function: variables that are captured across
/// function boundaries and need StoreContext/LoadContext semantics.
///
/// A binding is a context identifier if:
/// - It is reassigned from inside a nested function (`reassignedByInnerFn`), OR
/// - It is reassigned AND referenced from inside a nested function
///   (`reassigned && referencedByInnerFn`)
///
/// This is the Rust equivalent of the TypeScript `FindContextIdentifiers` pass.
pub fn find_context_identifiers(
    func: &FunctionNode<'_>,
    scope_info: &ScopeInfo,
) -> HashSet<BindingId> {
    let func_start = match func {
        FunctionNode::FunctionDeclaration(d) => d.base.start.unwrap_or(0),
        FunctionNode::FunctionExpression(e) => e.base.start.unwrap_or(0),
        FunctionNode::ArrowFunctionExpression(a) => a.base.start.unwrap_or(0),
    };
    let func_scope = scope_info
        .node_to_scope
        .get(&func_start)
        .copied()
        .unwrap_or(scope_info.program_scope);

    let mut visitor = ContextIdentifierVisitor {
        scope_info,
        function_stack: Vec::new(),
        binding_info: HashMap::new(),
    };
    let mut walker = AstWalker::with_initial_scope(scope_info, func_scope);

    // Walk params and body (like Babel's func.traverse())
    match func {
        FunctionNode::FunctionDeclaration(d) => {
            for param in &d.params {
                walker.walk_pattern(&mut visitor, param);
            }
            walker.walk_block_statement(&mut visitor, &d.body);
        }
        FunctionNode::FunctionExpression(e) => {
            for param in &e.params {
                walker.walk_pattern(&mut visitor, param);
            }
            walker.walk_block_statement(&mut visitor, &e.body);
        }
        FunctionNode::ArrowFunctionExpression(a) => {
            for param in &a.params {
                walker.walk_pattern(&mut visitor, param);
            }
            match a.body.as_ref() {
                ArrowFunctionBody::BlockStatement(block) => {
                    walker.walk_block_statement(&mut visitor, block);
                }
                ArrowFunctionBody::Expression(expr) => {
                    walker.walk_expression(&mut visitor, expr);
                }
            }
        }
    }

    // Supplement the walker-based analysis with referenceToBinding data.
    // The AST walker doesn't visit identifiers inside type annotations,
    // but Babel's traverse (used by TS findContextIdentifiers) does.
    // After scope extraction includes type annotation references,
    // we check if any reassigned binding has references in nested function scopes
    // via referenceToBinding — matching the TS behavior.
    //
    // We must skip declaration sites (e.g., the `x` in `function x() {}`),
    // which are included in reference_to_binding but are not true references.
    // Prefer node-ID comparison (immune to position-0 collisions from synthetic
    // nodes), falling back to position when node-IDs are unavailable.
    let declaration_ref_positions = build_declaration_positions(scope_info);
    for (&ref_pos, &binding_id) in &scope_info.reference_to_binding {
        let info = match visitor.binding_info.get(&binding_id) {
            Some(info) if info.reassigned && !info.referenced_by_inner_fn => info,
            _ => continue,
        };
        let _ = info;
        if declaration_ref_positions.contains(&(binding_id, ref_pos)) {
            continue;
        }
        let binding = &scope_info.bindings[binding_id.0 as usize];
        // Check if ref_pos is inside a nested function scope
        for (&scope_start, &scope_id) in &scope_info.node_to_scope {
            if scope_start <= ref_pos {
                if let Some(&scope_end) = scope_info.node_to_scope_end.get(&scope_start) {
                    if ref_pos < scope_end
                        && matches!(
                            scope_info.scopes[scope_id.0 as usize].kind,
                            ScopeKind::Function
                        )
                        && is_captured_by_function(scope_info, binding.scope, scope_id)
                    {
                        visitor
                            .binding_info
                            .get_mut(&binding_id)
                            .unwrap()
                            .referenced_by_inner_fn = true;
                        break;
                    }
                }
            }
        }
    }

    // Collect results
    visitor
        .binding_info
        .into_iter()
        .filter(|(_, info)| {
            info.reassigned_by_inner_fn || (info.reassigned && info.referenced_by_inner_fn)
        })
        .map(|(id, _)| id)
        .collect()
}
